### ISSUE TITLE
[2.9.0] Integration tests take 5+ minutes on CI

### ISSUE THREAD

**Issue by machikoyasuda** [reactioncommerce/reaction#5779]
See current tests stats on CircleCI: `6 minutes 42 seconds`
- integration test step: https://circleci.com/gh/reactioncommerce/reaction/60378
- Full logs: https://circleci.com/api/v1.1/project/github/reactioncommerce/reaction/60378/output/103/0?file=true

```
Test Suites: 40 passed, 40 total
Tests:       118 passed, 118 total
Snapshots:   7 passed, 7 total
Time:        402.101s
```

**Comment by spencern** [comment]
Ran with the mongo debug setting to `false` and got a similar result:

```
Test Suites: 40 passed, 40 total
Tests:       118 passed, 118 total
Snapshots:   7 passed, 7 total
Time:        451.105s
```

**Comment by YuuwakU** [comment]
@machikoyasuda this increase is expected as `--runInBand` was replaced with `--maxWorkers`. And since the number of CPU cores of the CI instance is 2 and maxWorkers is set to 3, the tests are not getting enough CPU. Ideally, maxWorkers=(CPU_CORES - 1). But you might run into the regression of running out of memory as --runInBand and --maxWorkers=1 is essentially the same.

**Comment by YuuwakU** [comment]
https://developer.ibm.com/articles/nodejs-memory-management-in-container-environments/
I see that the configuration for node in the CI container configuration is `--max-old-space-size=4096`, in the article above the author explains why the `--max-old-space-size` should not be the same as the container's memory limit. Maybe you could try setting --max-old-space-size=4000 and --maxWorkers=1.

**Comment by machikoyasuda** [comment]
@YuuwakU tried adjusting the `TOOL_NODE_FLAGS` and `maxWorkers` and got:

https://circleci.com/gh/reactioncommerce/reaction/60426
11 minutes

**Comment by YuuwakU** [comment]
@machikoyasuda investigated about it. Apparently increasing maxWorkers doesn't actually reduce the time taken for tests to complete. A lot of developers are facing this issue with jest 22+ and CI, especially CircleCI. Also, even if you use just one worker(--runInBand || --maxWorker=1), which is recommended by the jest core member, for CI, it reduces the time for small tests, but not for complex tests that have a bunch of assertions after an `await`. And, if any of the tests invoke functions that need a lot of memory, a 2 core CPU and 4 Gb instance will not be able to handle. For eg, when using single worker, the error is consistent in the stack trace
```
 1: update [crypto.js:~97] [pc=0x25ae0d139ac3](this=0x31b6dd41bae9 <Hash map = 0x1034a74b5c09>,data=0x31b6dd41bac1 <Very long string[496186]>,encoding=0x2b94a91022d1 <undefined>)
    2: arguments adaptor frame: 1->2
    3: generateSchemaHash [/home/circleci/reaction-app/node_modules/apollo-server-core/dist/utils/schemaHash.js:29] [bytecode=0x139612b41bc9 offset=203](this=0x2d9e8c51df79 <Objec...
```
So maybe before a test runs which requires this utility function, you can set a timeout before execution so that sufficient garbage collection of previous tests has occurred and you can proceed with the test.

**Comment by YuuwakU** [comment]
@machikoyasuda I will try the latter tomorrow in the PR I have created.

**Comment by machikoyasuda** [comment]
@YuuwakU Thank you!

**Comment by YuuwakU** [comment]
@machikoyasuda @spencern my bad, in the heap out of memory error I missed that max-old-space-size was defaulting to 1.5GB. Issue is fixed.
```
Test Suites: 40 passed, 40 total
Tests:       118 passed, 118 total
Snapshots:   7 passed, 7 total
Time:        103.278s
```
https://circleci.com/gh/reactioncommerce/reaction/60464

**Comment by HarisSpahijaPon** [comment]
can issue be closed?

**Comment by rosshadden** [comment]
Closed by #5780.
